call "add_to_journal" before "maybe_flush_pre_publish_cache" in pre_publish

tomyouyou opened the pull request:
To reproduce the issue:
1. pre_publish a delivered message.
2. Put its SeqId into delivered_cache, and might call "flush_pre_publish_cache" and "flush_delivered_cache", 
   then put its "del" data into the journal_entries of the corresponding segment.
3. Call add_to_journal and action_to_entry with a "?PUB" Action. 
4. in action_to_entry, "array:get(RelSeq, JEntries)" will return {no_pub,    del, no_ack}, but has none case for a "?PUB" Action, so an error will occur like "{{case_clause,{no_pub,del,no_ack}},[{rabbit_queue_index,action_to_entry,3..."

lukebakken commented:
@tomyouyou - if possible, could you provide a script that reproduces this issue by publishing messages a certain way? If not, could you provide the _exact_ Erlang code you're running to reproduce it?

tomyouyou commented:
@lukebakken
I use "PIKA" which is a kind of python rabbitmq client for test, No other scripts nor erlang code.

to reproduce the issue by the following steps:
1. Insert "{msg_store_io_batch_size, 2}" to /etc/rabbitmq/rabbitmq.config, 
    so "reduce_memory_use" could call "push_betas_to_deltas" much more easily.
2. retart broker, and run "rabbitmqctl set_vm_memory_high_watermark absolute 100M". 
    because the broker will consume about 65M memory when it is idle, and 65/100 > 0.5, 
    so the broker might call "reduce_memory_use" when processing any message.
3. declare a queue
4. create a consumer who never ack any message.  attach the consume to the  declared queue.
5. send some messages to the queue.
6. stop the consumer, so the broker will requeue all the unacked messages.
7. a crash might occur like "** {{case_clause,{no_pub,del,no_ack}},[{rabbit_queue_index,action_to_entry,3,[{file,"src/rabbit_queue_index.erl"},{line,838}]},{rabbit_queue_index,add_to_journal,3,[{file,"src/rabbit_queue_index.erl"},{line,814}]},{rabbit_queue_index,add_to_journal,3,[{file,"src/rabbit_queue_index.erl"},{line,805}]},{rabbit_variable_queue,maybe_batch_write_index_to_disk,3,[{file,"src/rabbit_variable_queue.erl"},{line,1992}]},{rabbit_variable_queue,push_betas_to_deltas1,4,[{file,"src/rabbit_variable_queue.erl"},{line,2817}]},{rabbit_variable_queue,push_betas_to_deltas,2,[{file,"src/rabbit_variable_queue.erl"},{line,2757}]},{rabbit_variable_queue,reduce_memory_use,1,[{file,"src/rabbit_variable_queue.erl"},{line,2493}]},{rabbit_variable_queue,set_ram_duration_target,2,[{file,"src/rabbit_variable_queue.erl"},{line,844}]}]}
"
8. it might need retry serval times to reproduce the crash. please follow the steps "2 => 3 ...=> 6" each time.

michaelklishin commented:
@tomyouyou thanks for the details.

lukebakken commented:
@tomyouyou thank you, I will try to reproduce this today.

lukebakken commented:
@tomyouyou if you could share the code you're using to reproduce this, it would be very helpful. I have not been able to do so yet.

tomyouyou commented:
@lukebakken
some additional descriptions:
1. the target of all the previous steps is to pre_publish delivered messages. 
    so BQ:batch_publish_delivered persistent messages could also reproduce the issue.
2. as for the "5" step "send some messages to the queue", the number of message should be greater than  
    "?SEGMENT_ENTRY_COUNT". If you could, you can adjust the "?SEGMENT_ENTRY_COUNT" to a smaller value, such as 3.
3. as for the "6" step, please wait some seconds to stop the consumer. because set_ram_duration_target need take time to cut down DurationTarget to trigger reduce_memory_use.

4. the following code is my consumer message handler function:
def callback_ack(ch, method, properties, body):
    ts = 2
    time.sleep(ts)
    ack = False
    #ack = True
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s" % (body[:20], ack, method.delivery_tag))
    else:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s" % (body, ack, method.delivery_tag))
    if ack:
        ch.basic_ack(delivery_tag = method.delivery_tag)

lukebakken commented:
Thanks for the additional information. Unfortunately, I still can't reproduce this on my laptop. Questions -

* How are you running RabbitMQ? What operating system, are you using a virtual machine or physical hardware, how much RAM, etc? My main interest is determining how fast your RabbitMQ server is.
* Are you running your producers / consumers on a different machine than RabbitMQ?
* Can you please share your ***complete*, *runnable*** code rather than a sample? If I am not using the same code as you there is a good chance you are doing something differently than me. For instance, I have no idea what size your message bodies are.

tomyouyou commented:
@lukebakken
%%%%%%%%%%%%%%%%%%%%%
information about my environment:
1. Use a LINUX PC for test with 4G memrory and Dual-Core 2.60GHz CPU.
2. I have adjusted the "?SEGMENT_ENTRY_COUNT" to 3.

%%%%%%%%%%%%%%%%%%%%% 
time and output of operations:
%%%%%%%%%%%%%%%%%%%%%
[root@controller2 ~]# rpm -qi rabbitmq-server
Name        : rabbitmq-server
Version     : 3.7.11
[root@controller2 erlang]# systemctl  restart rabbitmq-server
[root@controller2 erlang]# date;rabbitmqctl set_vm_memory_high_watermark absolute 100M
Sat Sep  7 10:14:24 CST 2019
Setting memory threshold on rabbit@controller2 to 100M bytes ...
[root@controller2 erlang]# date;python msg_test.py create_queue 192.168.1.66 'queue=q11;durable=false'
Sat Sep  7 10:14:31 CST 2019
connent to 192.168.1.66 ok
2019-09-07 10:14:31.824 create_queue q11
2019-09-07 10:14:31.827 to close connection
[root@controller2 erlang]# date;python msg_test.py create_exchange 192.168.1.66 'exchange=ex11;exchange_type=topic'
Sat Sep  7 10:14:31 CST 2019
connent to 192.168.1.66 ok
2019-09-07 10:14:32.034 create_exchange ex11 topic
2019-09-07 10:14:32.035 to close connection
[root@controller2 erlang]# date;python msg_test.py bind 192.168.1.66 'exchange=ex11;queue=q11;routing_key=exq11'
Sat Sep  7 10:14:33 CST 2019
connent to 192.168.1.66 ok
2019-09-07 10:14:33.297 bind ex11 q11 r_key:exq11
2019-09-07 10:14:33.299 to close connection
[root@controller2 erlang]# date;python msg_test.py publish 192.168.1.66 'exchange=ex11;routing_key=exq11;msg="_unique_id:1";count=100'
Sat Sep  7 10:14:52 CST 2019
connent to 192.168.1.66 ok
2019-09-07 10:14:52.226 publish ex11 r_key:exq11, 0
2019-09-07 10:14:52.227 publish ex11 r_key:exq11, 1
2019-09-07 10:14:52.227 publish ex11 r_key:exq11, 2
2019-09-07 10:14:52.228 publish ex11 r_key:exq11, 3
2019-09-07 10:14:52.228 publish ex11 r_key:exq11, 4
2019-09-07 10:14:52.228 publish ex11 r_key:exq11, 5
2019-09-07 10:14:52.229 publish ex11 r_key:exq11, 6
2019-09-07 10:14:52.229 publish ex11 r_key:exq11, 7
2019-09-07 10:14:52.229 publish ex11 r_key:exq11, 8
2019-09-07 10:14:52.230 publish ex11 r_key:exq11, 9
2019-09-07 10:14:52.230 publish ex11 r_key:exq11, 10
2019-09-07 10:14:52.230 publish ex11 r_key:exq11, 11
2019-09-07 10:14:52.231 publish ex11 r_key:exq11, 12
2019-09-07 10:14:52.231 publish ex11 r_key:exq11, 13
2019-09-07 10:14:52.231 publish ex11 r_key:exq11, 14
2019-09-07 10:14:52.232 publish ex11 r_key:exq11, 15
2019-09-07 10:14:52.232 publish ex11 r_key:exq11, 16
2019-09-07 10:14:52.232 publish ex11 r_key:exq11, 17
2019-09-07 10:14:52.233 publish ex11 r_key:exq11, 18
2019-09-07 10:14:52.233 publish ex11 r_key:exq11, 19
2019-09-07 10:14:52.233 publish ex11 r_key:exq11, 20
2019-09-07 10:14:52.234 publish ex11 r_key:exq11, 21
2019-09-07 10:14:52.234 publish ex11 r_key:exq11, 22
2019-09-07 10:14:52.234 publish ex11 r_key:exq11, 23
2019-09-07 10:14:52.235 publish ex11 r_key:exq11, 24
2019-09-07 10:14:52.235 publish ex11 r_key:exq11, 25
2019-09-07 10:14:52.235 publish ex11 r_key:exq11, 26
2019-09-07 10:14:52.236 publish ex11 r_key:exq11, 27
2019-09-07 10:14:52.236 publish ex11 r_key:exq11, 28
2019-09-07 10:14:52.236 publish ex11 r_key:exq11, 29
2019-09-07 10:14:52.237 publish ex11 r_key:exq11, 30
2019-09-07 10:14:52.237 publish ex11 r_key:exq11, 31
2019-09-07 10:14:52.237 publish ex11 r_key:exq11, 32
2019-09-07 10:14:52.238 publish ex11 r_key:exq11, 33
2019-09-07 10:14:52.238 publish ex11 r_key:exq11, 34
2019-09-07 10:14:52.238 publish ex11 r_key:exq11, 35
2019-09-07 10:14:52.239 publish ex11 r_key:exq11, 36
2019-09-07 10:14:52.239 publish ex11 r_key:exq11, 37
2019-09-07 10:14:52.239 publish ex11 r_key:exq11, 38
2019-09-07 10:14:52.240 publish ex11 r_key:exq11, 39
2019-09-07 10:14:52.240 publish ex11 r_key:exq11, 40
2019-09-07 10:14:52.240 publish ex11 r_key:exq11, 41
2019-09-07 10:14:52.241 publish ex11 r_key:exq11, 42
2019-09-07 10:14:52.241 publish ex11 r_key:exq11, 43
2019-09-07 10:14:52.241 publish ex11 r_key:exq11, 44
2019-09-07 10:14:52.241 publish ex11 r_key:exq11, 45
2019-09-07 10:14:52.242 publish ex11 r_key:exq11, 46
2019-09-07 10:14:52.242 publish ex11 r_key:exq11, 47
2019-09-07 10:14:52.242 publish ex11 r_key:exq11, 48
2019-09-07 10:14:52.243 publish ex11 r_key:exq11, 49
2019-09-07 10:14:52.243 publish ex11 r_key:exq11, 50
2019-09-07 10:14:52.243 publish ex11 r_key:exq11, 51
2019-09-07 10:14:52.244 publish ex11 r_key:exq11, 52
2019-09-07 10:14:52.244 publish ex11 r_key:exq11, 53
2019-09-07 10:14:52.244 publish ex11 r_key:exq11, 54
2019-09-07 10:14:52.245 publish ex11 r_key:exq11, 55
2019-09-07 10:14:52.245 publish ex11 r_key:exq11, 56
2019-09-07 10:14:52.245 publish ex11 r_key:exq11, 57
2019-09-07 10:14:52.246 publish ex11 r_key:exq11, 58
2019-09-07 10:14:52.246 publish ex11 r_key:exq11, 59
2019-09-07 10:14:52.246 publish ex11 r_key:exq11, 60
2019-09-07 10:14:52.247 publish ex11 r_key:exq11, 61
2019-09-07 10:14:52.247 publish ex11 r_key:exq11, 62
2019-09-07 10:14:52.247 publish ex11 r_key:exq11, 63
2019-09-07 10:14:52.248 publish ex11 r_key:exq11, 64
2019-09-07 10:14:52.248 publish ex11 r_key:exq11, 65
2019-09-07 10:14:52.248 publish ex11 r_key:exq11, 66
2019-09-07 10:14:52.248 publish ex11 r_key:exq11, 67
2019-09-07 10:14:52.249 publish ex11 r_key:exq11, 68
2019-09-07 10:14:52.249 publish ex11 r_key:exq11, 69
2019-09-07 10:14:52.249 publish ex11 r_key:exq11, 70
2019-09-07 10:14:52.250 publish ex11 r_key:exq11, 71
2019-09-07 10:14:52.250 publish ex11 r_key:exq11, 72
2019-09-07 10:14:52.250 publish ex11 r_key:exq11, 73
2019-09-07 10:14:52.251 publish ex11 r_key:exq11, 74
2019-09-07 10:14:52.251 publish ex11 r_key:exq11, 75
2019-09-07 10:14:52.251 publish ex11 r_key:exq11, 76
2019-09-07 10:14:52.252 publish ex11 r_key:exq11, 77
2019-09-07 10:14:52.252 publish ex11 r_key:exq11, 78
2019-09-07 10:14:52.252 publish ex11 r_key:exq11, 79
2019-09-07 10:14:52.253 publish ex11 r_key:exq11, 80
2019-09-07 10:14:52.253 publish ex11 r_key:exq11, 81
2019-09-07 10:14:52.253 publish ex11 r_key:exq11, 82
2019-09-07 10:14:52.254 publish ex11 r_key:exq11, 83
2019-09-07 10:14:52.254 publish ex11 r_key:exq11, 84
2019-09-07 10:14:52.254 publish ex11 r_key:exq11, 85
2019-09-07 10:14:52.255 publish ex11 r_key:exq11, 86
2019-09-07 10:14:52.255 publish ex11 r_key:exq11, 87
2019-09-07 10:14:52.255 publish ex11 r_key:exq11, 88
2019-09-07 10:14:52.255 publish ex11 r_key:exq11, 89
2019-09-07 10:14:52.256 publish ex11 r_key:exq11, 90
2019-09-07 10:14:52.256 publish ex11 r_key:exq11, 91
2019-09-07 10:14:52.256 publish ex11 r_key:exq11, 92
2019-09-07 10:14:52.257 publish ex11 r_key:exq11, 93
2019-09-07 10:14:52.257 publish ex11 r_key:exq11, 94
2019-09-07 10:14:52.257 publish ex11 r_key:exq11, 95
2019-09-07 10:14:52.258 publish ex11 r_key:exq11, 96
2019-09-07 10:14:52.258 publish ex11 r_key:exq11, 97
2019-09-07 10:14:52.258 publish ex11 r_key:exq11, 98
2019-09-07 10:14:52.259 publish ex11 r_key:exq11, 99
2019-09-07 10:14:53.260 to close connection
[root@controller2 erlang]#
[root@controller2 erlang]# date;python msg_test.py consume 192.168.1.66 'queue=q11;auto_ack=false'
Sat Sep  7 10:14:54 CST 2019
connent to 192.168.1.66 ok
2019-09-07 10:14:54.817  [*] Waiting for messages from q11. To exit press CTRL+C
2019-09-07 10:14:56.820  [x] Received '_unique_id:0 aaaaaaa' ack:False delivery_tag:1 sleep:2
2019-09-07 10:14:58.822  [x] Received '_unique_id:1 aaaaaaa' ack:False delivery_tag:2 sleep:2
2019-09-07 10:15:00.824  [x] Received '_unique_id:2 aaaaaaa' ack:False delivery_tag:3 sleep:2
2019-09-07 10:15:02.826  [x] Received '_unique_id:3 aaaaaaa' ack:False delivery_tag:4 sleep:2
2019-09-07 10:15:04.828  [x] Received '_unique_id:4 aaaaaaa' ack:False delivery_tag:5 sleep:2
2019-09-07 10:15:06.831  [x] Received '_unique_id:5 aaaaaaa' ack:False delivery_tag:6 sleep:2
2019-09-07 10:15:08.833  [x] Received '_unique_id:6 aaaaaaa' ack:False delivery_tag:7 sleep:2
2019-09-07 10:15:10.835  [x] Received '_unique_id:7 aaaaaaa' ack:False delivery_tag:8 sleep:2
2019-09-07 10:15:12.837  [x] Received '_unique_id:8 aaaaaaa' ack:False delivery_tag:9 sleep:2
2019-09-07 10:15:14.839  [x] Received '_unique_id:9 aaaaaaa' ack:False delivery_tag:10 sleep:2
2019-09-07 10:15:16.842  [x] Received '_unique_id:10 aaaaaa' ack:False delivery_tag:11 sleep:2
2019-09-07 10:15:18.844  [x] Received '_unique_id:11 aaaaaa' ack:False delivery_tag:12 sleep:2
2019-09-07 10:15:20.846  [x] Received '_unique_id:12 aaaaaa' ack:False delivery_tag:13 sleep:2
2019-09-07 10:15:22.848  [x] Received '_unique_id:13 aaaaaa' ack:False delivery_tag:14 sleep:2
2019-09-07 10:15:24.850  [x] Received '_unique_id:14 aaaaaa' ack:False delivery_tag:15 sleep:2
2019-09-07 10:15:26.851  [x] Received '_unique_id:15 aaaaaa' ack:False delivery_tag:16 sleep:2
2019-09-07 10:15:28.853  [x] Received '_unique_id:16 aaaaaa' ack:False delivery_tag:17 sleep:2
2019-09-07 10:15:30.856  [x] Received '_unique_id:17 aaaaaa' ack:False delivery_tag:18 sleep:2
2019-09-07 10:15:32.858  [x] Received '_unique_id:18 aaaaaa' ack:False delivery_tag:19 sleep:2
2019-09-07 10:15:34.858  [x] Received '_unique_id:19 aaaaaa' ack:False delivery_tag:20 sleep:2
2019-09-07 10:15:36.860  [x] Received '_unique_id:20 aaaaaa' ack:False delivery_tag:21 sleep:2
2019-09-07 10:15:38.862  [x] Received '_unique_id:21 aaaaaa' ack:False delivery_tag:22 sleep:2
^C2019-09-07 10:15:39.906 BaseException
2019-09-07 10:15:39.906 to close connection
[root@controller2 erlang]#

%%%%%%%%%%%%%%%%%%%%%
crash information in /var/log/rabbitmq/rabbit@controller2.log:
%%%%%%%%%%%%%%%%%%%%%
** Reason for termination ==
** {{case_clause,{no_pub,del,no_ack}},[{rabbit_queue_index,action_to_entry,3,[{file,"src/rabbit_queue_index.erl"},{line,838}]},{rabbit_queue_index,add_to_journal,3,[{file,"src/rabbit_queue_index.erl"},{line,814}]},{rabbit_queue_index,add_to_journal,3,[{file,"src/rabbit_queue_index.erl"},{line,805}]},{rabbit_variable_queue,maybe_batch_write_index_to_disk,3,[{file,"src/rabbit_variable_queue.erl"},{line,1992}]},{rabbit_variable_queue,push_betas_to_deltas1,4,[{file,"src/rabbit_variable_queue.erl"},{line,2817}]},{rabbit_variable_queue,push_betas_to_deltas,2,[{file,"src/rabbit_variable_queue.erl"},{line,2757}]},{rabbit_variable_queue,reduce_memory_use,1,[{file,"src/rabbit_variable_queue.erl"},{line,2493}]},{rabbit_variable_queue,set_ram_duration_target,2,[{file,"src/rabbit_variable_queue.erl"},{line,844}]}]}
2019-09-07 10:15:41.868 [error] <0.410.0> CRASH REPORT Process <0.410.0> with 0 neighbours exited with reason: no case clause matching {no_pub,del,no_ack} in rabbit_queue_index:action_to_entry/3 line 838 in gen_server2:terminate/3 line 1166
2019-09-07 10:15:41.870 [error] <0.409.0> Supervisor {<0.409.0>,rabbit_amqqueue_sup} had child rabbit_amqqueue started with rabbit_prequeue:start_link({amqqueue,{resource,<<"/">>,queue,<<"q11">>},false,false,none,[],none,[],[],[],undefined,undefined,...}, declare, <0.408.0>) at <0.410.0> exit with reason no case clause matching {no_pub,del,no_ack} in rabbit_queue_index:action_to_entry/3 line 838 in context child_terminated
2019-09-07 10:15:41.874 [critical] <0.477.0> DBG [init_dirty:620] CleanShutdown:false

%%%%%%%%%%%%%%%%%%%%%
msg_test.py
%%%%%%%%%%%%%%%%%%%%%
#!/usr/bin/env python
import time
import sys
import atexit
import pika
import logging

import pdb
#pdb.set_trace()


_Log = logging.getLogger()
#logging.basicConfig(level=logging.DEBUG)

sys.argv.pop(0)

if len(sys.argv) < 3:
    usage()
    sys.exit(0)
    
cmd = sys.argv.pop(0)
hostname = sys.argv.pop(0)
connection = pika.BlockingConnection(pika.ConnectionParameters(host=hostname,port=5672,virtual_host='/',\
credentials=pika.credentials.PlainCredentials('guest', 'guest')))
channel = connection.channel()

print("connent to %s ok" % hostname)

def is_true(value):
    if not value:
        return False
    if value.isdigit() and int(value) > 0:
        return True
    if value == "true" or value == "yes" or value == "on":
        return True
    return False

@atexit.register
def stop():
    print get_time_stamp() + "to close connection"
    connection.close()
    
def callback(ch, method, properties, body):
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body[:20], method.delivery_tag))
    else:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body, method.delivery_tag))
def callback_ack(ch, method, properties, body):
    ts = 2
    time.sleep(ts)
    ack = False
    #ack = True
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body[:20], ack, method.delivery_tag, ts))
    else:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body, ack, method.delivery_tag, ts))
    if ack:
        ch.basic_ack(delivery_tag = method.delivery_tag)    
    
exchange_name = ''
ex_type = None
queue_name = None
r_key = None
msg=None
auto_ack_l = True    
auto_delete_l = False
real_call_back = callback
count=1
msg_size = 100
prefetch = 0
allchs = False
durable_l = False
persistent_l = False
rate=1000
options = sys.argv.pop(0).split(';')
#print options
#pdb.set_trace()
def get_time_stamp():
    ct = time.time()
    local_time = time.localtime(ct)
    data_head = time.strftime("%Y-%m-%d %H:%M:%S", local_time)
    data_secs = (ct - int(ct)) * 1000
    time_stamp = "%s.%03d " % (data_head, data_secs)
    return time_stamp

for opt in options:
    tmps = opt.split('=')
    if len(tmps) != 2:
        print("invalid options:%s" % options)
        sys.exit(0)
    if tmps[0] == "exchange":
        exchange_name = tmps[1]
    elif tmps[0] == "exchange_type":
        ex_type = tmps[1]
    elif tmps[0] == "queue":
        queue_name = tmps[1]
    elif tmps[0] == "routing_key":
        r_key = tmps[1]
    elif tmps[0] == "msg":
        msg = tmps[1]
    elif tmps[0] == "auto_delete":
        auto_delete_l = is_true(tmps[1])
    elif tmps[0] == "auto_ack":
        auto_ack_l = is_true(tmps[1])
        if not auto_ack_l:
            real_call_back = callback_ack
    elif tmps[0] == "count":
        count = int(tmps[1])
    elif tmps[0] == "prefetch":
        prefetch = int(tmps[1])
    elif tmps[0] == "all_channels":
        allchs = is_true(tmps[1])
    elif tmps[0] == "durable":
        durable_l = is_true(tmps[1])
    elif tmps[0] == "persistent":
        persistent_l = is_true(tmps[1])
    elif tmps[0] == "msg_size":
        msg_size = int(tmps[1])
    elif tmps[0] == "rate":
        rate = int(tmps[1])
try:
    if cmd == "create_queue" and queue_name:   
        print get_time_stamp() + "%s %s" % (cmd, queue_name)
        channel.queue_declare(queue=queue_name, durable = durable_l)
    elif cmd == "create_exchange" and exchange_name and ex_type: 
        print get_time_stamp() + "%s %s %s" % (cmd, exchange_name, ex_type)
        channel.exchange_declare(exchange=exchange_name,
                             exchange_type=ex_type, auto_delete=auto_delete_l,durable = durable_l)
    elif cmd == "delete_exchange" and exchange_name: 
        print get_time_stamp() + "%s %s" % (cmd, exchange_name)
        channel.exchange_delete(exchange=exchange_name)
    elif cmd == "bind" and exchange_name and queue_name:  
        print get_time_stamp() + "%s %s %s r_key:%s" % (cmd, exchange_name, queue_name, r_key)
        channel.queue_bind(exchange=exchange_name,
                       queue=queue_name,
                       routing_key=r_key)
    elif cmd == "publish" and exchange_name is not None and r_key is not None and msg: 
        properties_l=None
        if persistent_l:
            properties_l=pika.BasicProperties(
                         delivery_mode = 2, # make message persistent
                      )
        msg_plus = 'a' * msg_size
        cnt1 = count/rate
        if cnt1 == 0:
            cnt1 = 1
            rate = count
        ji = 0
        for j in range(0,cnt1):
            for i in range(0, rate):
                print get_time_stamp() + "%s %s r_key:%s, %d" % (cmd, exchange_name, r_key, ji)
                ji += 1
                msg2send="_unique_id:%d %s" % (i, msg_plus)
                channel.basic_publish(exchange=exchange_name, 
                              routing_key=r_key, 
                              body=msg2send,
                              properties=properties_l)
            time.sleep(1)         
    elif cmd == "consume" and queue_name: 
        if prefetch > 0:
            pdb.set_trace()
            channel.basic_qos(prefetch_count=prefetch,all_channels=allchs)
        #channel.basic_consume(queue_name, real_call_back, auto_ack = auto_ack_l)  
        channel.basic_consume(real_call_back, queue_name, no_ack = auto_ack_l)
        print(get_time_stamp() + ' [*] Waiting for messages from %s. To exit press CTRL+C' % queue_name)
        channel.start_consuming()
        print(get_time_stamp() + '!!!quit consuing')
        print(get_time_stamp() + 'to exit')
    else:
        usage()
except BaseException:
    print get_time_stamp() + "BaseException"
    pass

lukebakken commented:
@tomyouyou I'll try again to reproduce this but honestly without the code you're using it will probably be unlikely. If there is any way you can share it that would be ideal. Thanks again.

tomyouyou commented:
@lukebakken  the following is the complete code msg_test.py 
#!/usr/bin/env python
import time
import sys
import atexit
import pika
import logging

import pdb
#pdb.set_trace()


_Log = logging.getLogger()
#logging.basicConfig(level=logging.DEBUG)

sys.argv.pop(0)

if len(sys.argv) < 3:
    usage()
    sys.exit(0)
    
cmd = sys.argv.pop(0)
hostname = sys.argv.pop(0)
connection = pika.BlockingConnection(pika.ConnectionParameters(host=hostname,port=5672,virtual_host='/',\
credentials=pika.credentials.PlainCredentials('guest', 'guest')))
channel = connection.channel()

print("connent to %s ok" % hostname)

def is_true(value):
    if not value:
        return False
    if value.isdigit() and int(value) > 0:
        return True
    if value == "true" or value == "yes" or value == "on":
        return True
    return False

@atexit.register
def stop():
    print get_time_stamp() + "to close connection"
    connection.close()
    
def callback(ch, method, properties, body):
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body[:20], method.delivery_tag))
    else:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body, method.delivery_tag))
def callback_ack(ch, method, properties, body):
    ts = 0
    if ts > 0:
        time.sleep(ts)
    ack = False
    ack = True
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body[:20], ack, method.delivery_tag, ts))
    else:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body, ack, method.delivery_tag, ts))
    if ack:
        ch.basic_ack(delivery_tag = method.delivery_tag)    
    
exchange_name = ''
ex_type = None
queue_name = None
r_key = None
msg=None
auto_ack_l = True    
auto_delete_l = False
real_call_back = callback
count=1
msg_size = 100
prefetch = 0
allchs = False
durable_l = False
persistent_l = False
rate=1000
options = sys.argv.pop(0).split(';')
#print options
#pdb.set_trace()

def get_time_stamp():
    tt = time.time()
    hms = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(tt))
    return "%s.%03d " % (hms, (tt - int(tt)) * 1000)

for opt in options:
    tmps = opt.split('=')
    if len(tmps) != 2:
        print("invalid options:%s" % options)
        sys.exit(0)
    if tmps[0] == "exchange":
        exchange_name = tmps[1]
    elif tmps[0] == "exchange_type":
        ex_type = tmps[1]
    elif tmps[0] == "queue":
        queue_name = tmps[1]
    elif tmps[0] == "routing_key":
        r_key = tmps[1]
    elif tmps[0] == "msg":
        msg = tmps[1]
    elif tmps[0] == "auto_delete":
        auto_delete_l = is_true(tmps[1])
    elif tmps[0] == "auto_ack":
        auto_ack_l = is_true(tmps[1])
        if not auto_ack_l:
            real_call_back = callback_ack
    elif tmps[0] == "count":
        count = int(tmps[1])
    elif tmps[0] == "prefetch":
        prefetch = int(tmps[1])
    elif tmps[0] == "all_channels":
        allchs = is_true(tmps[1])
    elif tmps[0] == "durable":
        durable_l = is_true(tmps[1])
    elif tmps[0] == "persistent":
        persistent_l = is_true(tmps[1])
    elif tmps[0] == "msg_size":
        msg_size = int(tmps[1])
    elif tmps[0] == "rate":
        rate = int(tmps[1])
try:
    if cmd == "create_queue" and queue_name:   
        print get_time_stamp() + "%s %s" % (cmd, queue_name)
        channel.queue_declare(queue=queue_name, durable = durable_l)
    elif cmd == "create_exchange" and exchange_name and ex_type: 
        print get_time_stamp() + "%s %s %s" % (cmd, exchange_name, ex_type)
        channel.exchange_declare(exchange=exchange_name,
                             exchange_type=ex_type, auto_delete=auto_delete_l,durable = durable_l)
    elif cmd == "delete_exchange" and exchange_name: 
        print get_time_stamp() + "%s %s" % (cmd, exchange_name)
        channel.exchange_delete(exchange=exchange_name)
    elif cmd == "bind" and exchange_name and queue_name:  
        print get_time_stamp() + "%s %s %s r_key:%s" % (cmd, exchange_name, queue_name, r_key)
        channel.queue_bind(exchange=exchange_name,
                       queue=queue_name,
                       routing_key=r_key)
    elif cmd == "publish" and exchange_name is not None and r_key is not None and msg: 
        properties_l=None
        if persistent_l:
            properties_l=pika.BasicProperties(
                         delivery_mode = 2, # make message persistent
                      )
        msg_plus = 'a' * msg_size
        cnt1 = count/rate
        if cnt1 == 0:
            cnt1 = 1
            rate = count
        ji = 0
        for j in range(0,cnt1):
            for i in range(0, rate):
                print get_time_stamp() + "%s %s r_key:%s, %d" % (cmd, exchange_name, r_key, ji)
                ji += 1
                msg2send="_unique_id:%d %s" % (i, msg_plus)
                channel.basic_publish(exchange=exchange_name, 
                              routing_key=r_key, 
                              body=msg2send,
                              properties=properties_l)
            time.sleep(1)         
    elif cmd == "consume" and queue_name: 
        if prefetch > 0:
            pdb.set_trace()
            channel.basic_qos(prefetch_count=prefetch,all_channels=allchs)
        #channel.basic_consume(queue_name, real_call_back, auto_ack = auto_ack_l)  
        channel.basic_consume(real_call_back, queue_name, no_ack = auto_ack_l)
        print(get_time_stamp() + ' [*] Waiting for messages from %s. To exit press CTRL+C' % queue_name)
        channel.start_consuming()
        print(get_time_stamp() + '!!!quit consuing')
        print(get_time_stamp() + 'to exit')
    else:
        usage()
except BaseException:
    print get_time_stamp() + "BaseException"
    pass

tomyouyou commented:
@lukebakken  the following is the complete code msg_test.py
``` python
#!/usr/bin/env python
import time
import sys
import atexit
import pika
import logging

import pdb
#pdb.set_trace()


_Log = logging.getLogger()
#logging.basicConfig(level=logging.DEBUG)

sys.argv.pop(0)

if len(sys.argv) < 3:
    usage()
    sys.exit(0)
    
cmd = sys.argv.pop(0)
hostname = sys.argv.pop(0)
connection = pika.BlockingConnection(pika.ConnectionParameters(host=hostname,port=5672,virtual_host='/',\
credentials=pika.credentials.PlainCredentials('guest', 'guest')))
channel = connection.channel()

print("connent to %s ok" % hostname)

def is_true(value):
    if not value:
        return False
    if value.isdigit() and int(value) > 0:
        return True
    if value == "true" or value == "yes" or value == "on":
        return True
    return False

@atexit.register
def stop():
    print get_time_stamp() + "to close connection"
    connection.close()
    
def callback(ch, method, properties, body):
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body[:20], method.delivery_tag))
    else:
        print(get_time_stamp() + " [x] Received %r delivery_tag:%s" % (body, method.delivery_tag))
def callback_ack(ch, method, properties, body):
    ts = 0
    if ts > 0:
        time.sleep(ts)
    ack = False
    ack = True
    if len(body) > 20:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body[:20], ack, method.delivery_tag, ts))
    else:
        print(get_time_stamp() + " [x] Received %r ack:%s delivery_tag:%s sleep:%s" % (body, ack, method.delivery_tag, ts))
    if ack:
        ch.basic_ack(delivery_tag = method.delivery_tag)    
    
exchange_name = ''
ex_type = None
queue_name = None
r_key = None
msg=None
auto_ack_l = True    
auto_delete_l = False
real_call_back = callback
count=1
msg_size = 100
prefetch = 0
allchs = False
durable_l = False
persistent_l = False
rate=1000
options = sys.argv.pop(0).split(';')
#print options
#pdb.set_trace()

def get_time_stamp():
    tt = time.time()
    hms = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(tt))
    return "%s.%03d " % (hms, (tt - int(tt)) * 1000)

for opt in options:
    tmps = opt.split('=')
    if len(tmps) != 2:
        print("invalid options:%s" % options)
        sys.exit(0)
    if tmps[0] == "exchange":
        exchange_name = tmps[1]
    elif tmps[0] == "exchange_type":
        ex_type = tmps[1]
    elif tmps[0] == "queue":
        queue_name = tmps[1]
    elif tmps[0] == "routing_key":
        r_key = tmps[1]
    elif tmps[0] == "msg":
        msg = tmps[1]
    elif tmps[0] == "auto_delete":
        auto_delete_l = is_true(tmps[1])
    elif tmps[0] == "auto_ack":
        auto_ack_l = is_true(tmps[1])
        if not auto_ack_l:
            real_call_back = callback_ack
    elif tmps[0] == "count":
        count = int(tmps[1])
    elif tmps[0] == "prefetch":
        prefetch = int(tmps[1])
    elif tmps[0] == "all_channels":
        allchs = is_true(tmps[1])
    elif tmps[0] == "durable":
        durable_l = is_true(tmps[1])
    elif tmps[0] == "persistent":
        persistent_l = is_true(tmps[1])
    elif tmps[0] == "msg_size":
        msg_size = int(tmps[1])
    elif tmps[0] == "rate":
        rate = int(tmps[1])
try:
    if cmd == "create_queue" and queue_name:   
        print get_time_stamp() + "%s %s" % (cmd, queue_name)
        channel.queue_declare(queue=queue_name, durable = durable_l)
    elif cmd == "create_exchange" and exchange_name and ex_type: 
        print get_time_stamp() + "%s %s %s" % (cmd, exchange_name, ex_type)
        channel.exchange_declare(exchange=exchange_name,
                             exchange_type=ex_type, auto_delete=auto_delete_l,durable = durable_l)
    elif cmd == "delete_exchange" and exchange_name: 
        print get_time_stamp() + "%s %s" % (cmd, exchange_name)
        channel.exchange_delete(exchange=exchange_name)
    elif cmd == "bind" and exchange_name and queue_name:  
        print get_time_stamp() + "%s %s %s r_key:%s" % (cmd, exchange_name, queue_name, r_key)
        channel.queue_bind(exchange=exchange_name,
                       queue=queue_name,
                       routing_key=r_key)
    elif cmd == "publish" and exchange_name is not None and r_key is not None and msg: 
        properties_l=None
        if persistent_l:
            properties_l=pika.BasicProperties(
                         delivery_mode = 2, # make message persistent
                      )
        msg_plus = 'a' * msg_size
        cnt1 = count/rate
        if cnt1 == 0:
            cnt1 = 1
            rate = count
        ji = 0
        for j in range(0,cnt1):
            for i in range(0, rate):
                print get_time_stamp() + "%s %s r_key:%s, %d" % (cmd, exchange_name, r_key, ji)
                ji += 1
                msg2send="_unique_id:%d %s" % (i, msg_plus)
                channel.basic_publish(exchange=exchange_name, 
                              routing_key=r_key, 
                              body=msg2send,
                              properties=properties_l)
            time.sleep(1)         
    elif cmd == "consume" and queue_name: 
        if prefetch > 0:
            pdb.set_trace()
            channel.basic_qos(prefetch_count=prefetch,all_channels=allchs)
        #channel.basic_consume(queue_name, real_call_back, auto_ack = auto_ack_l)  
        channel.basic_consume(real_call_back, queue_name, no_ack = auto_ack_l)
        print(get_time_stamp() + ' [*] Waiting for messages from %s. To exit press CTRL+C' % queue_name)
        channel.start_consuming()
        print(get_time_stamp() + '!!!quit consuing')
        print(get_time_stamp() + 'to exit')
    else:
        usage()
except BaseException:
    print get_time_stamp() + "BaseException"
    pass

```

lukebakken commented:
@tomyouyou - thank you for providing your code. In the future, you can save your code in a `.txt` file and attach it to a comment rather than pasting the entire output.

Also, please note that the code you provided does not run exactly as shown as the output in [this comment](https://github.com/rabbitmq/rabbitmq-server/pull/2096#issuecomment-529064302). The output shows a 2 second sleep in the `callback_ack` method as well as other subtle differences in the code you provided. I figured out your intent, and you can find the code I used [in this gist](https://gist.github.com/lukebakken/05cdfba78b57901c20d872c070151b2a#file-msg_test-py).

The reason we need to have _exactly_ the same code you are using is that it speeds up reproduction of your issues, and merging your pull requests. Now that I have code that works, I can reproduce your issue -

```
** Reason for termination == 
** {{case_clause,{no_pub,del,no_ack}},[{rabbit_queue_index,action_to_entry,3,[{file,"src/rabbit_queue_index.erl"},{line,813}]},
```

Thanks.

Again, in the future, it's important to provide code that reproduces your issue, and instructions for how to run that code.

lukebakken commented:
I verified that the fix in 5e8d4f1ee resolves this issue.

Backported to `v3.7.x`
